src/freenet/node/diagnostics/diagnostics/Diagnostics.java

```diff
@@ -0,0 +1,8 @@
+/* This code is part of Freenet. It is distributed under the GNU General
+ * Public License, version 2 (or at your option any later version). See
+ * http://www.gnu.org/ for further details of the GPL. */
+package freenet.node.diagnostics.diagnostics;
+
+public interface Diagnostics {
```

Comment:
I feel this name is somehow wrong because a `start` method and diagnostics have very little to do with each other…

Reply:
I don’t like this interface here at all.

I’d much rather have an interface for the `ThreadDiagnostics` which exposes the single `getThreads()` method. This way we are completely free to implement this however we see fit; exposing a concrete class will only increase the public API of Fred.

Actually, I would prefer to even expand this to the `NodeDiagnostics`  as well. Turn that into an interface with a single `getThreadDiagnostics()` method. During startup,`Node` knows how to create a, say, `DefaultNodeDiagnostics` object (with `Ticker` ’n stuff), and `DefaultNodeDiagnostics` knows how to create the `DefaultThreadDiagnostics` and also knows that it needs to be started. (Actually, both should maybe also be stoppable, now that I think of it? Can be left for a later PR, though, as long as a ticket for it is created. 🙂)

Yes, this increases the amount of seemingly useless code but at this point I really prefer to add as little as possible to Fred’s public API, and interfaces (together with some default implementation that should never be used without their respective interfaces) fit that bill perfectly.

Reply:
Seems reasonable. I actually don't understand the DefaultNodeDiagnostics part but I get the idea of having a simpler interface with only NodeDiagnostics providing access to the (possible) various diagnostics we can put into it.

Re: being able to stop them is something I was considering, so yeah, makes sense.

Reply:
If `NodeDiagnostics` is turned into an interface you need an implementation somewhere. The real use of the interface is that we are free to change implementations without having to change Fred’s public API. However, you still *need* at least one concrete implementation (which is created during the Node’s startup) and I simply suggest naming that `DefaultNodeDiagnostics` because it’s our default implementation. Any other name (that is *not* `NodeDiagnosticsImpl`) would be fine, too. 🙂